handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java

```diff
@@ -396,7 +409,42 @@ private int readEncryptedData(final ByteBuffer dst, final int pending) {
             }
         }
 
-        return 0;
+        return bioRead;
+    }
+
+    private SSLEngineResult readPendingBytesFromBIO(ByteBuffer dst, int bytesConsumed, int bytesProduced) {
+        // Check to see if the engine wrote data into the network BIO
+        int pendingNet = SSL.pendingWrittenBytesInBIO(networkBIO);
+        if (pendingNet > 0) {
+
+            // Do we have enough room in dst to write encrypted data?
+            int capacity = dst.remaining();
+            if (capacity < pendingNet) {
+                return new SSLEngineResult(
+                        BUFFER_OVERFLOW, getHandshakeStatus(pendingNet), bytesConsumed, bytesProduced);
+            }
+
+            // Write the pending data from the network BIO into the dst buffer
+            int produced = readEncryptedData(dst, pendingNet);
+
+            if (produced <= 0) {
+                // TODO: Use BIO_should_retry() to find the exact cause
```

Comment:
Lets update this comment and instead indicate why we do not use `BIO_should_retry`.  And also consider adding logging incase our assumptions are wrong.